src/topology/constructions.lean

```diff
@@ -1055,6 +1055,23 @@ lemma is_open_set_pi {i : set ι} {s : Πa, set (π a)} (hi : i.finite) (hs : 
   is_open (pi i s) :=
 by rw [pi_def]; exact (is_open_bInter hi $ assume a ha, (hs _ ha).preimage (continuous_apply _))
 
+lemma is_open_pi_iff [finite ι] {s : set (Πa, π a)} :
+  is_open s ↔
+  (∀ f, f ∈ s → ∃ (u : Πa, set (π a)), (∀ a, is_open (u a) ∧ f a ∈ u a) ∧ set.univ.pi u ⊆ s) :=
```

Comment:
Is there a version of this that doesn't need the `finite` assumption, and instead has a slightly different RHS?

Reply:
I added
```lean
lemma is_open_pi_iff {s : set (Πa, π a)} :
   is_open s ↔
   (∀ f, f ∈ s → ∃ (I : finset ι) (u : Π a, set (π a)),
   (∀ a, ∃ (t : set (π a)), t ⊆ u a ∧ is_open t ∧ f a ∈ t) ∧ (I : set ι).pi u ⊆ s) 
```
and rename the finite version to `is_open_pi_iff'`

Reply:
Although, there is probably a better statement for this result. Let me see...

Reply:
Yes, a simpler statement is:
```lean
lemma is_open_pi_iff {s : set (Πa, π a)} :
  is_open s ↔
  (∀ f, f ∈ s → ∃ (I : finset ι) (u : Π a, set (π a)),
  (∀ a, a ∈ I → is_open (u a) ∧ f a ∈ u a) ∧ (I : set ι).pi u ⊆ s)
```
but the proof I have for the moment is quite ugly so I'll have to work on it some more before I can push the new version.

Reply:
Here is the new version of `is_open_pi_iff` with its proof. I am not really happy with the proof and the fact that the proof of the finite version `is_open_pi_iff'` is simpler if I do it directly rather than using the general version `is_open_pi_iff`... But that's the best I could do.

Reply:
Looks better than anything I could come up with. Let's let @fpvandoorn take a final look.

Reply:
I think this can be formulated as "sets of the form `(I : set ι).pi u` form a filter basis for `\nhds x`". Maybe this can be formulated in the same was as [`filter.has_basis.prod_nhds'`](https://leanprover-community.github.io/mathlib_docs/topology/constructions.html#filter.has_basis.prod_nhds')?

The Zulip thread already contains mentions of filter bases, so I don't know if they were intentionally avoided.

Reply:
I guess what you have in mind is proving a finite version of [filter.has_basis_pi](https://leanprover-community.github.io/mathlib_docs/order/filter/pi.html#filter.has_basis_pi) and use that to prove `is_open_pi_iff'` which is the result that I actually need? I am not quite familiar with filter basis but I can give it a shot.

Reply:
No, I was thinking to do it without the assumption that `\io` is finite, so working with finite subsets of it. In particular, I think this is the lemma in the style of `filter.has_basis.prod_nhds'`:
```lean
import topology.constructions

open_locale topology
open set
variables {ι : Type*} {π ι' : ι → Type*} [∀ i, topological_space (π i)]

lemma filter.has_basis.pi_nhds {x : Π i, π i} {p : Π i, ι' i → Prop} {s : Π i, ι' i → set (π i)}
  (h : ∀ i, (𝓝 (x i)).has_basis (p i) (s i)) : 
  (𝓝 x).has_basis (λ I : finset ι × Π i, ι' i, ∀ i ∈ I.1, p i (I.2 i)) 
  (λ I, (I.1 : set ι).pi (λ i, s i (I.2 i))) :=
sorry
```

Reply:
But this requires characterizing neighborhoods in pi-types, and maybe that's not even easier than this.

Let me just merge this now as-is, it looks good. If you feel motivated, feel free to still to prove the above lemma in a follow-up PR.

bors merge